File: backend/src/main/java/com/climaticbeats/service/WeatherService.java
package com.climaticbeats.service;

import com.climaticbeats.dto.WeatherResponse;
import com.climaticbeats.entity.SearchHistory;
import com.climaticbeats.entity.User;
import com.climaticbeats.repository.SearchHistoryRepository;
import com.climaticbeats.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.Map;

@Service
public class WeatherService {

    @Autowired
    private WebClient.Builder webClientBuilder;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private SearchHistoryRepository searchHistoryRepository;

    @Value("${weather.api.key}")
    private String apiKey;

    @Value("${weather.api.url}")
    private String apiUrl;

    @Transactional
    public WeatherResponse getWeather(String city, String email) {
        if (apiKey == null || apiKey.isEmpty()) {
            throw new RuntimeException("Weather API key is not configured");
        }

        String url = apiUrl + "?q=" + city + "&appid=" + apiKey + "&units=metric";

        Map<String, Object> response = webClientBuilder.build()
                .get()
                .uri(url)
                .retrieve()
                .bodyToMono(Map.class)
                .block();

        if (response == null) {
            throw new RuntimeException("Failed to fetch weather data");
        }

        Map<String, Object> main = (Map<String, Object>) response.get("main");
        Map<String, Object> weather = ((java.util.List<Map<String, Object>>) response.get("weather")).get(0);
        Map<String, Object> wind = (Map<String, Object>) response.get("wind");
        Map<String, Object> sys = (Map<String, Object>) response.get("sys");

        WeatherResponse weatherResponse = new WeatherResponse();
        weatherResponse.setCity((String) response.get("name"));
        weatherResponse.setCountry((String) sys.get("country"));
        weatherResponse.setCondition((String) weather.get("main"));
        weatherResponse.setDescription((String) weather.get("description"));
        weatherResponse.setTemperature(((Number) main.get("temp")).doubleValue());
        weatherResponse.setFeelsLike(((Number) main.get("feels_like")).doubleValue());
        weatherResponse.setHumidity(((Number) main.get("humidity")).intValue());
        weatherResponse.setWindSpeed(wind.get("speed") != null ? ((Number) wind.get("speed")).doubleValue() : 0.0);
        weatherResponse.setIcon((String) weather.get("icon"));

        // Save search history
        if (email != null) {
            User user = userRepository.findByEmail(email)
                    .orElse(null);
            if (user != null) {
                SearchHistory searchHistory = new SearchHistory();
                searchHistory.setUser(user);
                searchHistory.setCity(weatherResponse.getCity());
                searchHistory.setCountryCode(weatherResponse.getCountry());
                searchHistoryRepository.save(searchHistory);
            }
        }

        return weatherResponse;
    }
}

File: backend/src/weather/weather.service.ts
import { Injectable } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { firstValueFrom } from 'rxjs';
import { WeatherResponseDto } from './dto/weather-response.dto';
import { SearchHistory } from '../history/entities/search-history.entity';
import { User } from '../auth/entities/user.entity';

@Injectable()
export class WeatherService {
  private baseUrl = process.env.OPEN_WEATHER_BASE_URL;
  private apiKey = process.env.OPEN_WEATHER_API_KEY;

  constructor(
    private httpService: HttpService,
    @InjectRepository(SearchHistory)
    private searchHistoryRepository: Repository<SearchHistory>,
  ) {}

  async getWeather(
    city: string,
    user?: User,
  ): Promise<WeatherResponseDto> {
    if (!this.apiKey) {
      throw new Error('OpenWeather API key not configured');
    }

    try {
      const url = `${this.baseUrl}/weather?q=${city}&appid=${this.apiKey}&units=metric`;
      const response = await firstValueFrom(this.httpService.get(url));
      const data = response.data;

      const weatherResponse: WeatherResponseDto = {
        city: data.name,
        country: data.sys.country,
        condition: data.weather[0].main,
        description: data.weather[0].description,
        temperature: data.main.temp,
        feelsLike: data.main.feels_like,
        humidity: data.main.humidity,
        windSpeed: data.wind.speed,
        icon: data.weather[0].icon,
      };

      // Save to search history if user is provided
      if (user) {
        const searchHistory = this.searchHistoryRepository.create({
          user,
          city: weatherResponse.city,
          countryCode: weatherResponse.country,
        });
        await this.searchHistoryRepository.save(searchHistory);
      }

      return weatherResponse;
    } catch (error) {
      throw new Error(`Failed to fetch weather: ${error.message}`);
    }
  }
}
